### ISSUE TITLE
Crash on NSData+ImageContentType.m:13 - +sd_contentTypeForImageData:

### ISSUE THREAD

**Issue by takasek** [rs/SDWebImage#1169]
Hi, I got two crashes with v 3.7.2.

----

**NSInvalidArgumentException** in **NSData(ImageContentType)**
-[__NSCFString getBytes:length:]: unrecognized selector sent to instance 0x156c9df0

## Stacktrace

    Pods/SDWebImage/SDWebImage/NSData+ImageContentType.m:13 - +[NSData(ImageContentType) sd_contentTypeForImageData:]
    Pods/SDWebImage/SDWebImage/UIImage+MultiFormat.m:22 - +[UIImage(MultiFormat) sd_imageWithData:]
    Pods/SDWebImage/SDWebImage/SDImageCache.m:270 - -[SDImageCache diskImageForKey:]
    Pods/SDWebImage/SDWebImage/SDImageCache.m:310 - _42-[SDImageCache queryDiskCacheForKey:done:]_block_invoke

----

**EXC_BAD_ACCESS** in **NSData(ImageContentType)**
Attempted to dereference garbage pointer 0xe302000.

## Stacktrace

    Pods/SDWebImage/SDWebImage/NSData+ImageContentType.m:13 - +[NSData(ImageContentType) sd_contentTypeForImageData:]
    Pods/SDWebImage/SDWebImage/UIImage+MultiFormat.m:22 - +[UIImage(MultiFormat) sd_imageWithData:]
    Pods/SDWebImage/SDWebImage/SDWebImageDownloaderOperation.m:368 - -[SDWebImageDownloaderOperation connectionDidFinishLoading:]
    Pods/SDWebImage/SDWebImage/SDWebImageDownloaderOperation.m:110 - -[SDWebImageDownloaderOperation start]

**Comment by bpoplauschi** [comment]
Could you share a demo project to easily replicate those crashes?

**Comment by takasek** [comment]
Sorry, they were reported by Bugsnag and I cannot figure out how to reproduce them.

**Comment by bpoplauschi** [comment]
Do you have any statistics on how many users run into those crashes? I can't do much without a clear repro scenario. Just by looking at the stacktrace, I assume it's a memory management issue since instead of an NSData, that method `sd_contentTypeForImageData:` receives an `NSCFString` object. This could have a lot of causes

**Comment by anton-matosov** [comment]
I have received similar crash today:
```
Crashed: NSOperationQueue 0x1b184140 :: NSOperation 0x16213d80 (QOS: USER_INTERACTIVE)
EXC_BAD_ACCESS KERN_INVALID_ADDRESS 0x000000000fd62000 
```

Stacktrace:
```
Thread : Crashed: NSOperationQueue 0x1b184140 :: NSOperation 0x16213d80 (QOS: USER_INTERACTIVE)
0  libsystem_platform.dylib       0x394fd990 _platform_memmove + 143
1  Foundation                     0x28187a2b -[NSData(NSData) getBytes:length:] + 118
2  Foundation                     0x28187a2b -[NSData(NSData) getBytes:length:] + 118
3  <AppName>                      0x1f0d11 +[NSData(ImageContentType) sd_contentTypeForImageData:] (NSData+ImageContentType.m:13)
4  <AppName>                      0x1fe06f +[UIImage(MultiFormat) sd_imageWithData:] (UIImage+MultiFormat.m:22)
5  <AppName>                      0x1f7327 -[SDWebImageDownloaderOperation connectionDidFinishLoading:] (SDWebImageDownloaderOperation.m:347)
6  CFNetwork                      0x26db4f69 __65-[NSURLConnectionInternal _withConnectionAndDelegate:onlyActive:]_block_invoke + 56
7  CFNetwork                      0x26db4f21 -[NSURLConnectionInternal _withConnectionAndDelegate:onlyActive:] + 184
8  CFNetwork                      0x26db5015 -[NSURLConnectionInternal _withActiveConnectionAndDelegate:] + 36
9  CFNetwork                      0x26c74e57 ___ZN27URLConnectionClient_Classic26_delegate_didFinishLoadingEU13block_pointerFvvE_block_invoke + 78
10 CFNetwork                      0x26d2e937 ___ZN27URLConnectionClient_Classic18_withDelegateAsyncEPKcU13block_pointerFvP16_CFURLConnectionPK33CFURLConnectionClientCurrent_VMaxE_block_invoke_2 + 70
11 libdispatch.dylib              0x3936cec3 _dispatch_client_callout + 22
12 libdispatch.dylib              0x39374713 _dispatch_block_invoke + 450
13 CFNetwork                      0x26c65e9b RunloopBlockContext::_invoke_block(void const*, void*) + 18
14 CoreFoundation                 0x27366c29 CFArrayApplyFunction + 36
15 CFNetwork                      0x26c65d83 RunloopBlockContext::perform() + 182
16 CFNetwork                      0x26c65c4d MultiplexerSource::perform() + 216
17 CFNetwork                      0x26c65ae1 MultiplexerSource::_perform(void*) + 48
18 CoreFoundation                 0x27416827 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 14
19 CoreFoundation                 0x27416417 __CFRunLoopDoSources0 + 454
20 CoreFoundation                 0x2741477f __CFRunLoopRun + 806
21 CoreFoundation                 0x273671e9 CFRunLoopRunSpecific + 516
22 CoreFoundation                 0x273a7fab CFRunLoopRun + 98
23 <AppName>                      0x1f5fa3 -[SDWebImageDownloaderOperation start] (SDWebImageDownloaderOperation.m:107)
24 Foundation                     0x28212cc5 __NSOQSchedule_f + 192
25 libdispatch.dylib              0x3937704f _dispatch_queue_drain + 1762
26 libdispatch.dylib              0x3937002f _dispatch_queue_invoke + 282
27 libdispatch.dylib              0x393783dd _dispatch_root_queue_drain + 400
28 libdispatch.dylib              0x3937824b _dispatch_worker_thread3 + 94
29 libsystem_pthread.dylib        0x39501e29 _pthread_wqthread + 1024
```

**Comment by bpoplauschi** [comment]
Let's close this issue, see if it happens again with 4.0.0 (4.0.0-beta is available for download, 4.0.0 official release is on its way).
